shared/libraries/motor.cc

```diff
@@ -534,4 +539,149 @@ void SteeringMotor::UpdateData(const uint8_t data[]) {
+  uint8_t data[8] = {0};
+  data[0] = 0xff;
+  data[1] = 0xff;
+  data[2] = 0xff;
+  data[3] = 0xff;
+  data[4] = 0xff;
+  data[5] = 0xff;
+  data[6] = 0xff;
+  data[7] = 0xfe;
+  motor->can_->Transmit(motor->tx_id_, data, 8);
+}
+
+void Motor4310::SetOutput(float position, float velocity, float kp, float kd, float torque) {
+  kp_set_ = kp;
+  kd_set_ = kd;
+  pos_set_ = position;
+  vel_set_ = velocity;
+  torque_set_ = torque;
+}
+
+void Motor4310::SetOutput(float position, float velocity) {
+  pos_set_ = position;
+  vel_set_ = velocity;
+}
+
+void Motor4310::SetOutput(float velocity) {
+  vel_set_ = velocity;
+}
+
+void Motor4310::TransmitOutput(Motor4310* motor) {
+  uint8_t data[8] = {0};
+  int16_t kp_tmp, kd_tmp, pos_tmp, vel_tmp, torque_tmp;
+
+  // converting float to unsigned int before transmitting
+  kp_tmp = float_to_uint(kp_set_, 0.0, 500.0, 12);
+  kd_tmp = float_to_uint(kd_set_, 0, 5, 12);
+  pos_tmp = float_to_uint(pos_set_, -12.5, 12.5, 16);
+  vel_tmp = float_to_uint(vel_set_, -45.0, 45.0, 12);
+  torque_tmp = float_to_uint(torque_set_, -18, 18, 12);
+
+  if (mode_ == MIT){
+    data[0] = pos_tmp >> 8;
+    data[1] = pos_tmp & 0x00ff;
+    data[2] = (vel_tmp >> 4) & 0x00ff;
+    data[3] = ((vel_tmp & 0x000f) << 4) | ((kp_tmp >> 8) & 0x000f);
+    data[4] = kp_tmp & 0x00ff;
+    data[5] = (kd_tmp >> 4) & 0x00ff;
+    data[6] = ((kd_tmp & 0x000f) << 4) | ((torque_tmp >> 8) & 0x000f);
+    data[7] = torque_tmp & 0x00ff;
+  } else if (mode_ == POS_VEL) {
+    uint8_t *pbuf, *vbuf;
+    pbuf = (uint8_t*) &pos_set_;
+    vbuf = (uint8_t*) &vel_set_;
+    data[0] = *pbuf;
+    data[1] = *(pbuf + 1);
+    data[2] = *(pbuf + 2);
+    data[3] = *(pbuf + 3);
+    data[4] = *vbuf;
+    data[5] = *(vbuf + 1);
+    data[6] = *(vbuf + 2);
+    data[7] = *(vbuf + 3);
+  } else if (mode_ == VEL) {
+    uint8_t  *vbuf;
+    vbuf = (uint8_t*) &vel_set_;
+    data[0] = *vbuf;
+    data[1] = *(vbuf + 1);
+    data[2] = *(vbuf + 2);
+    data[3] = *(vbuf + 3);
+  } else {
+    RM_EXPECT_TRUE(false, "Invalid mode number!");
+  }
+  motor->can_->Transmit(motor->tx_id_, data, 8);
+  connection_flag_ = true;  // temp
+}
+
+void Motor4310::UpdateData(const uint8_t data[]) {
+  // TODO need test
+  raw_pos_ = data[1]<<8 | data[2];
+  raw_vel_ = data[3]<<4 | data[4]>>4;
+  raw_torque_ = data[4] - ((data[4]>>4)<<4);
+  raw_torque_ = raw_torque_<<8 | data[5];
+  raw_mosTemp_ = data[6];
+  raw_rotorTemp_ = data[7];
+
+  connection_flag_ = true;
+}
+
+void Motor4310::PrintData() {
+  set_cursor(0, 0);
+  clear_screen();
+  print("Position: % .4f \r\n", raw_pos_);
+  print("Velocity: % .4f \r\n", raw_vel_);
+  print("Torque: % .4f \r\n", raw_torque_);
+  print("Rotor temp: % .4f \r\n", raw_rotorTemp_);
+}
+
+int16_t Motor4310::float_to_uint(float x, float x_min, float x_max, int bits) {
+  float span = x_max - x_min;
+  float offset = x_min;
+  return (int16_t) ((x-offset) * ((float)((1<<bits)-1))/span);
+}
+
```

Comment:
Suggest moving this to utils / static function instead of making it a member function.